File: 0003.无重复字符的最长子串/Solution.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;

        int start = 0, end = 0, ans = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        while (end < s.length()) {
            if (map.containsKey(s.charAt(end)) && map.get(s.charAt(end)) >= start) {
                start = map.get(s.charAt(end)) + 1;
            }
            map.put(s.charAt(end), end);
            ans = Math.max(ans, end - start + 1);
            end++;
        }
        return ans;
    }
}

File: 0003.无重复字符的最长子串/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start, end, ans = 0, 0, 0
        kv = dict()

        while end < len(s):
            if s[end] in kv and kv[s[end]] >= start:
                start = kv[s[end]] + 1

            kv[s[end]] = end
            ans = max(ans, end - start + 1)
            end += 1

        return ans
